frontends/p4/typeChecking/typeChecker.h

```diff
@@ -80,23 +76,25 @@ bool hasVarbitsOrUnions(const TypeMap *typeMap, const IR::Type *type);
 // In fact, several passes do modify the program such that types are invalidated.
 // For example, enum elimination converts enum values into integers.  After such
 // changes the typemap has to be cleared and types must be recomputed from scratch.
-class TypeInference : public Transform {
-    // Input: reference map
-    ReferenceMap *refMap;
+class TypeInference : public Transform, public ResolutionContext {
     // Output: type map
     TypeMap *typeMap;
     const IR::Node *initialNode;
+    std::shared_ptr<MinimalNameGenerator> nameGen;
+
+    TypeInference(TypeMap *typeMap, std::shared_ptr<MinimalNameGenerator> nameGen);
 
  public:
     // @param readOnly If true it will assert that it behaves like
     //        an Inspector.
-    TypeInference(ReferenceMap *refMap, TypeMap *typeMap, bool readOnly = false,
-                  bool checkArrays = true);
+    explicit TypeInference(TypeMap *typeMap, bool readOnly = false, bool checkArrays = true,
```

Comment:
I am thinking we should make these structs instead of a list of booleans. Can be easy to get them wrong even with the `

I really like option structs for that: https://abseil.io/tips/173

Reply:
Yeah, these are very nice. I think there were some neat wrappers somewhere to add some sugar on top of that.